File: dynamic_programming/rod_cutting.py
# Recursive top-down implementation
def cut_rod(p: list, n: int):
    if n == 0:
        return 0
    q = -1
    for i in range(n):
        q = max(q, p[i] + cut_rod(p, n - i - 1))
    return q
# Using dynamic programming for optimal rod cutting
def cut_rod_optimal(p: list, n: int):
    r = [0 for _ in range(len(p))]
    def memoized_cut_rod_aux(p, n, r):
        if r[n-1] > 0:
            return r[n-1]
        q = 0
        for i in range(n):
            q = max(q, p[i] + memoized_cut_rod_aux(p, n - i - 1, r))
        r[n-1] = q
        return q
    return memoized_cut_rod_aux(p, n, r)
# Bottom-up cut rod
def bottom_up_cut_rod(p: list, n: int):
    r = [0 for _ in range(len(p) + 1)]
    for j in range(0, n):
        q = -1
        for i in range(0, j + 1):
            q = max(q, p[i] + r[j-i])
        r[j + 1] = q
    print(r)
    return r[n]

if __name__ == "__main__":
    p = [1, 5, 8, 9, 10, 17, 17, 20, 24, 30]
    print("Recursive top-down implementation: {}".format(cut_rod(p, 3)))
    print("Memorized optimal implementation: {}".format(cut_rod_optimal(p, 3)))
    print("Bottom-up implementation: {}".format(bottom_up_cut_rod(p, 3)))

File: dynamic_programming/rod_cutting.go
package main

import (
	"fmt"
	"math"
)

func cut_rod(p []int64, n int64) int64 {
	if n == 0 {
		return 0
	}
	var q int64
	var i int64
	for ; i < n; i++ {
		q = int64(math.Max(float64(q), float64(p[i]+cut_rod(p, n-i-1))))
	}
	return q
}

func rut_rod_optimal(p []int64, n int64) int64 {
	r := make([]int64, len(p))
	return memorized_cut_rod_aux(p, n, r)
}

func memorized_cut_rod_aux(p []int64, n int64, r []int64) int64 {
	if n <= 0 {
		return 0
	}
	if r[n-1] > 0 {
		return r[n-1]
	}
	var q int64
	var i int64
	i = 0
	for ; i < n; i++ {
		q = int64(math.Max(float64(q), float64(p[i]+memorized_cut_rod_aux(p, n-i-1, r))))
	}
	r[n-1] = q
	return q
}

func bottom_up_cut_rod(p []int64, n int64) int64 {
	r := make([]int64, len(p)+1)
	var i int64
	for ; i < n; i++ {
		q := int64(0)
		for j := int64(0); j <= i; j++ {
			q = int64(math.Max(float64(q), float64(p[j]+r[i-j])))
		}
		r[i+1] = q
	}
	return r[n]
}

func main() {
	p := []int64{1, 5, 8, 9, 10, 17, 17, 20, 24, 30}
	fmt.Println(cut_rod(p, 3))
	fmt.Println(rut_rod_optimal(p, 3))
	fmt.Println(bottom_up_cut_rod(p, 3))
}
